test_embark/staking.js

```diff
@@ -0,0 +1,193 @@
+    await token.approve(stakingAddress, defaultAmount).send()
+    // stake tokens
+    await staking.stakeFor(other, defaultAmount, web3.utils.asciiToHex('')).send()
+
+    const finalOwnerBalance = parseInt((await token.balanceOf(owner).call()).valueOf(), 10)
+    const finalOtherBalance = parseInt((await token.balanceOf(other).call()).valueOf(), 10)
+    const finalStakingBalance = parseInt((await token.balanceOf(stakingAddress).call()).valueOf(), 10)
+    assert.equal(finalOwnerBalance, initialOwnerBalance - defaultAmount, "owner balance should match")
+    assert.equal(finalOtherBalance, initialOtherBalance, "other balance should match")
+    assert.equal(finalStakingBalance, initialStakingBalance + defaultAmount, "Staking app balance should match")
+    assert.equal((await staking.totalStakedFor(owner).call()).valueOf(), 0, "staked value for owner should match")
+    assert.equal((await staking.totalStakedFor(other).call()).valueOf(), defaultAmount, "staked value for other should match")
+  })
+
+  it('unstakes', async () => {
+    const initialOwnerBalance = parseInt((await token.balanceOf(owner).call()).valueOf(), 10)
+    const initialStakingBalance = parseInt((await token.balanceOf(stakingAddress).call()).valueOf(), 10)
+
+    await approveAndStake()
+
+    // unstake half of them
+    await staking.unstake(defaultAmount / 2, web3.utils.asciiToHex('')).send()
+
+    const finalOwnerBalance = parseInt((await token.balanceOf(owner).call()).valueOf(), 10)
+    const finalStakingBalance = parseInt((await token.balanceOf(stakingAddress).call()).valueOf(), 10)
+    assert.equal(finalOwnerBalance, initialOwnerBalance - defaultAmount / 2, "owner balance should match")
+    assert.equal(finalStakingBalance, initialStakingBalance + defaultAmount / 2, "Staking app balance should match")
+    assert.equal((await staking.totalStakedFor(owner).call()).valueOf(), defaultAmount / 2, "staked value should match")
+  })
+
+  it('fails unstaking 0 amount', async () => {
+    await approveAndStake()
+    return assertRevert(async () => {
+      await staking.unstake(0, web3.utils.asciiToHex('')).send()
+    })
+  })
+
+  it('fails unstaking more than staked', async () => {
+    await approveAndStake()
+    return assertRevert(async () => {
+      await staking.unstake(defaultAmount + 1, web3.utils.asciiToHex('')).send()
+    })
+  })
+
+  context('History', async () => {
+    it('supports history', async () => {
+      assert.equal(await staking.supportsHistory().call(), true, "It should support History")
+    })
+
+    it('has correct "last staked for"', async () => {
+      const blockNumber = await staking.getBlockNumber64Ext().call()
+      const lastStaked = blockNumber + 5
+      await staking.setBlockNumber64(lastStaked).send()
+      await approveAndStake()
+      assert.equal(await staking.lastStakedFor(owner).call(), lastStaked, "Last staked for should match")
+    })
+
+    it('has correct "total staked for at"', async () => {
```

Comment:
What are we really testing here? We should be making multiple stakes and then check that at all checkpoints the staked amount is what it should be